gen/nodes.ts

```diff
@@ -14,7 +14,7 @@ export interface SourceLocation {
 }
 
 export interface Node extends Printable {
-  type: string;
+  type: "Node";
```

Comment:
> Not all nodes have `type`, just buildable ones. So, the generated types aren't accurate there.

Do you mean that non-buildable types should have `type: string` like before, since (for example) `"Node"` is not a type that any parser would produce? Would that cause any problems for the type narrowing behavior?

Reply:
Yes, exactly. The types felt misleading in that case / not true to the underlying runtime behavior. For example, we'd never have a node with the type "Node" at runtime.

Just added a commit https://github.com/benjamn/ast-types/pull/300/commits/2d12dcb938c72ee8b1863334e91bb250dfe3bcf3 that changed it back to using the type's "own" fields to make the generated types more accurate.

Then, followed that up with a commit https://github.com/benjamn/ast-types/pull/300/commits/02f7b86b276b1f813aaaa1913e60943d0eb681dd that adds an export for a union type of all buildable types. These are the only ones that actually would have been constructed at runtime.

The type narrowing works great at the top-level.
![image](https://user-images.githubusercontent.com/723547/50853436-14868500-1337-11e9-8964-41d68070cf4c.png)

The issue is if any "kind" (as in gen/kinds) unions something that introduces `type: string` the narrowing breaks.
For example, `CatchClause` has `param: K.PatternKind | null;`.
`type PatternKind = N.Pattern | N.Expression | ...`
Where `export interface Pattern extends Node {}` and `export interface Node extends Printable { type: string; ... }`
![image](https://user-images.githubusercontent.com/723547/50853961-a6db5880-1338-11e9-8f9c-feb5757cbdc1.png)

Here is a self-contained demonstration of this behavior [on the TypeScript Playground](http://www.typescriptlang.org/play/index.html#src=type%20Omit%3CT%2C%20K%3E%20%3D%20Pick%3CT%2C%20Exclude%3Ckeyof%20T%2C%20K%3E%3E%3B%0D%0A%0D%0Ainterface%20A%20%7B%0D%0A%20%20%20%20type%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20B%20extends%20A%20%7B%0D%0A%20%20%20%20type%3A%20%22B%22%3B%0D%0A%20%20%20%20b%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20C%20extends%20A%20%7B%0D%0A%20%20%20%20type%3A%20%22C%22%3B%0D%0A%20%20%20%20c%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20D%20extends%20A%2C%20Omit%3CB%2C%20%22type%22%3E%20%7B%0D%0A%20%20%20%20type%3A%20%22D%22%3B%0D%0A%20%20%20%20d%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Atype%20U1%20%3D%20A%20%7C%20B%20%7C%20C%20%7C%20D%3B%0D%0Atype%20U2%20%3D%20B%20%7C%20C%20%7C%20D%3B%0D%0A%0D%0Afunction%20fn1(u%3A%20U1)%20%7B%0D%0A%20%20%20%20if%20(u.type%20%3D%3D%3D%20%22C%22)%20%7B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20type%20does%20not%20narrow%20because%20of%20%60type%3A%20string%60%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Afunction%20fn2(u%3A%20U2)%20%7B%0D%0A%20%20%20%20if%20(u.type%20%3D%3D%3D%20%22B%22)%20%7B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20type%20narrows%20to%20%60B%60%0D%0A%20%20%20%20%20%20%20%20console.log(u.b)%3B%0D%0A%20%20%20%20%7D%20else%20if%20(u.type%20%3D%3D%3D%20%22C%22)%20%7B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20type%20narrows%20to%20%60C%60%0D%0A%20%20%20%20%20%20%20%20console.log(u.c)%3B%0D%0A%20%20%20%20%7D%20else%20if%20(u.type%20%3D%3D%3D%20%22D%22)%20%7B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20type%20narrows%20to%20%60D%60%0D%0A%20%20%20%20%20%20%20%20console.log(u.b%2C%20u.d)%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A).

I noticed that the `@babel/types` uses the union-ed node type's `type` field - that might have been added to address this situation. https://gist.github.com/brieb/e312c8f2565c4b2cde5996e1f648780e#file-index-d-ts-L41

So, if we change the `Node` interface to have `type: ASTNode["type"]` we can retain the string literal union / ability to narrow.
![image](https://user-images.githubusercontent.com/723547/50854283-7c3dcf80-1339-11e9-9ad6-38f4f95c2d70.png)

So, I added another commit to use `type: ASTNode["type"]` rather than `type: string` https://github.com/benjamn/ast-types/pull/300/commits/493d64fb0515e238a253de4bb2854247fa9a79ea

Reply:
I take that last part back about being able to narrow the nested `CatchClause` `param`. Using `type: ASTNode["type"]` doesn't actually work as I'd expected. Going to fix that up and push a commit.

```
function exampleFn(node: ASTNode) {
  if (node.type === "CatchClause") {
    if (node.param) {
      if (node.param.type === "BlockStatement") {
        // Oops. `node.param` is still `CatchClause.param: Expression | Pattern | XMLAnyName | XMLQualifiedIdentifier | XMLFunctionQualifiedIdentifier | XMLAttributeSelector | XMLFilterExpression | XMLElement | XMLList`
      }
    }
  }
}
```